### ISSUE TITLE
[Bug] Inline Temporary Variable Refactoring produces invalid code (VS2017 RTM)

### ISSUE THREAD

**Issue by axel-habermaier** [dotnet/roslyn#17746]
**Version Used**: 
VS 2017 RTM

**Steps to Reproduce**:

1. Add the following code to a C# project:
```csharp
class X
{
	void M<T>(T o) // where T : class does not make a difference
	{
		var x = o as X;
		if (x != null)
			;
	}
}
```
2. Invoke the "Inline Temporary Variable" refactoring on variable ``x``.

**Actual Behavior**:
The refactoring preview shows that the generated code is going to be invalid. The refactoring can still be applied, yielding
```csharp
class X
{
	void M<T>(T o)
	{
		if (o is X x)
			;
	}
}
```

The following compiler error results:
```error CS8121: An expression of type T cannot be handled by a pattern of type X.```

**Expected Behavior**:
Ideally, the compiler would support the code generated by the refactoring, as I understand the is pattern match in this case to be semantically equivalent to the original code.

At the very least, though, the refactoring should not be offered by the IDE.

**Comment by alrz** [comment]
This was reported before and is per spec as `(X)o` is not permitted.  Perhaps the analyzer should check convertibility of the expression in the same sense of direct cast to cover corner-cases until this is fixed in the compiler (if possible).

**Comment by alrz** [comment]
Related: https://github.com/dotnet/csharplang/issues/154

**Comment by axel-habermaier** [comment]
Thanks @alrz. A very unfortunate language limitation, especially given that the refactoring is offered.
